torch/_functorch/aot_autograd.py

```diff
@@ -2772,6 +3038,277 @@ def traced_forward(*primals_fwd_seed_fwd_base_offset):
         PhiloxStateTracker.record_state(fwd_seed, fwd_base_offset, "forward")
         return traced_forward, (*args, fwd_seed, fwd_base_offset)
 
+
+# This function takes in a pytree of arguments and unwraps any tensor subclasses.
+# Annoyingly, we can't use pytrees to perform the unwrapping, because unwrapping returns
+# a list of tensors that we would then need to concat together.
+# Instead, we specialize the logic for the inference vs. joint graph case.
+def unwrap_tensor_subclasses(wrapped_args, *, trace_joint: bool):
+    if trace_joint:
+        assert isinstance(wrapped_args, tuple) and len(wrapped_args) == 2
+        assert isinstance(wrapped_args[0], (tuple, list)) and isinstance(wrapped_args[1], (tuple, list))
+        unwrapped_args_fw = []
+        for a in wrapped_args[0]:
+            if isinstance(a, torch.Tensor) and is_traceable_wrapper_subclass(a):
+                attrs, _ = a.__tensor_flatten__()
+                unwrapped_args_fw += [getattr(a, attr) for attr in attrs]
+            else:
+                unwrapped_args_fw += [a]
+        unwrapped_args_tangents = []
+        for a in wrapped_args[1]:
+            if isinstance(a, torch.Tensor) and is_traceable_wrapper_subclass(a):
+                attrs, _ = a.__tensor_flatten__()
+                unwrapped_args_tangents += [getattr(a, attr) for attr in attrs]
+            else:
+                unwrapped_args_tangents += [a]
+        unwrapped_args = (unwrapped_args_fw, unwrapped_args_tangents)
+    else:
+        unwrapped_args_fw = []
+        assert isinstance(wrapped_args, (list, tuple))
+        for a in wrapped_args:
+            if isinstance(a, torch.Tensor) and is_traceable_wrapper_subclass(a):
+                attrs, _ = a.__tensor_flatten__()
+                unwrapped_args_fw += [getattr(a, attr) for attr in attrs]
+            else:
+                unwrapped_args_fw += [a]
+        unwrapped_args = unwrapped_args_fw
+    return unwrapped_args
+
+# Turns a flattened list of tensor arguments into (maybe) subclass tensors.
+# This function is used both at trace time and runtime, so we have an is_runtime flag telling us which context we're in.
+def wrap_tensor_subclasses_simple(
```

Comment:
what makes this simple?

Reply:
Renamed to `wrap_tensor_subclasses` (and renamed the other fn `wrap_tensor_subclasses_maybe_joint`)